### PR TITLE
Better typing for exception variables

### PR DESCRIPTION
**Pull request by jez**
### Summary of new errors

There were two classes of errors that I saw when fixing errors in prep for landing this PR:

- Pinning errors that looked like this:

  ```ruby
  begin
    raises_a
  rescue A => e
  end
  
  2.times do
    begin
      puts(e.class)
      raises_b
    rescue B => e
    end
  end
  ```

  This is a pinning error, because `A` and `B` are different exception
  types. The result of `puts(e.class)` would change between the first and
  second iterations of the loop (from `A` to `B`).
  
  This situation is relatively rare (happens < 100 times in all of
  Stripe's codebase), and the fix is pretty easy (just pick a different
  name for the variable).

- Errors where Sorbet correctly detects that `e` is uninitialized after the `begin`/`rescue` blocks ends (because all exceptional control flow paths also unconditionally raised or returned).

  These indicate logic errors and should be fixed or silenced like normal.

### Motivation

This does not yet completely fix the problem of typing exceptions, but it makes them better in a few ways:

- It does not handle clauses that have `ensure`, due to some weird limitations in Sorbet's control flow handling.
- We no longer consider `e` to be `T.untyped` after the `begin`/`rescue` block ends. Now, it's always typed as `T.nilable(StandardError)` (or something more specific if the rescue clause was more specific.
- Due to the way things shook out in inference, the loc that we used to respond to an LSP hover query with always matched the `<get-current-exception>` type before the `e` type when hovering over the `=> e`. Now you can hover over `=> e` and get the right type.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_OIhPia3mcXCptC
→ https://go/builds/bui_OIhPdKSEpYQCsB
→ https://go/builds/bui_OIhPWFVemb4C9e

**Comment by froydnj** [comment]
This is the same change as #6937 ?

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Ogg4UsMXCfuIr1
→ https://go/builds/bui_Ogg4GtnwOExEDA
→ https://go/builds/bui_Ogg4Ky83RRWcBf

File: test/testdata/cfg/rescue_complex.rb [hunk 1]
```diff
@@ -42,7 +42,7 @@ def multiple_rescue_classes_varuse()
       baz
     end
 
-    T.reveal_type(baz) # error: Revealed type: `T.untyped`
+    T.reveal_type(baz) # error: Revealed type: `T.nilable(T.any(LoadError, SocketError))`
```
Comment (froydnj) [inline test/testdata/cfg/rescue_complex.rb:45]: Yay!

**Review by froydnj — APPROVED** [review]

**Pull request merged**